### ISSUE TITLE
[C#] Provide a mechanism to include custom JsonSerializerSettings, or return raw JSON from Decrypt

### ISSUE THREAD

**Issue by rleverence-gd** [godaddy/asherah#316]
When retrieving data from a JSON blob, Newtonsoft.Json attempts to parse any data that looks like dates internally as a DateTime. If you attempt to turn one of these parsed date properties into a string, you'll end up with a date of an arbitrary format and potentially not the string you expect. This happens at parse/deserialization time, and there's no way to change the token type of the property post-parse. The only way to prevent this from happening is to provide custom JsonDeserializerSettings with `DateParseHandling` set to `None`. Even setting the default JSON deserializer settings globally via `JsonConvert.DefaultSettings` won't solve this issue.

https://github.com/godaddy/asherah/blob/db905abee1f91f770c2d497433105fadbcc1a6d8/csharp/AppEncryption/AppEncryption/SessionJsonImpl.cs#L31

The solution I'd like to see is to either provide a way to allow a user to provide custom `JsonDeserializerSettings` as an optional parameter to `Decrypt`, or include a method to return the raw JSON data.

**Comment by sushantmimani** [comment]
Hi @rleverence-gd. Can you provide an example we can use to reproduce this issue?

**Comment by rleverence-gd** [comment]
Sure thing. I have a dotnetfiddle with an example here: https://dotnetfiddle.net/CfNhQ8

```c#
using System;
using GoDaddy.Asherah.AppEncryption;
using GoDaddy.Asherah.AppEncryption.Util;
using Newtonsoft.Json.Linq;

public class Thing
{
	public string ArbitraryString { get; set; }
}

public class Program
{
	public static void Main()
	{
		var thing = new Thing
		{
			ArbitraryString = DateTime.UtcNow.ToString("o")
		};
		thing.ArbitraryString.Dump();
		
		var thingJObject = JObject.FromObject(thing);
		
		var sessionFactory = SessionFactory.NewBuilder("productId", "serviceId")
			.WithInMemoryMetastore()
			.WithNeverExpiredCryptoPolicy()
			.WithStaticKeyManagementService("staticMasterKey!")
			.Build();
		
		using (var sessionJson = sessionFactory.GetSessionJson("partitionId"))
		{
			var encryptedPayload = sessionJson.Encrypt(thingJObject);
			var decryptedJObject = sessionJson.Decrypt(encryptedPayload);
			var decodedThing = decryptedJObject.ToObject<Thing>();
			decodedThing.ArbitraryString.Dump();
		}	
	}
}
```

If you run it, you'll see the result below:
```
Dumping object(String)
 2020-10-13T16:32:01.0730790Z
Dumping object(String)
 10/13/2020 16:32:01
```

The expectation would be that the results would be the same--they're strings, after all. But Newtonsoft.Json wants to try and parse it as a date time, and then when that value is pulled again as a string, it uses its own default formatting to make the value different from how it originally was.

**Comment by sushantmimani** [comment]
Hey @rleverence-gd 

I ran the code snippet with an additional print statement to see what the `Decrypt` method is returning. It looks like the return value is what you'd expect.
```c#
using System;
using GoDaddy.Asherah.AppEncryption;
using GoDaddy.Asherah.AppEncryption.Util;
using Newtonsoft.Json.Linq;

public class Thing
{
	public string ArbitraryString { get; set; }
}

public class Program
{
	public static void Main()
	{
		var thing = new Thing
		{
			ArbitraryString = DateTime.UtcNow.ToString("o")
		};
		thing.ArbitraryString.Dump();
		
		var thingJObject = JObject.FromObject(thing);
		Console.Write("\nPayload: " + thingJObject);
		
		var sessionFactory = SessionFactory.NewBuilder("productId", "serviceId")
			.WithInMemoryMetastore()
			.WithNeverExpiredCryptoPolicy()
			.WithStaticKeyManagementService("staticMasterKey!")
			.Build();
		
		using (var sessionJson = sessionFactory.GetSessionJson("partitionId"))
		{
			var encryptedPayload = sessionJson.Encrypt(thingJObject);
			var decryptedJObject = sessionJson.Decrypt(encryptedPayload);
			Console.WriteLine("\nDecrypted payload: " + decryptedJObject);
			var decodedThing = decryptedJObject.ToObject<Thing>();
			decodedThing.ArbitraryString.Dump();
		}	
	}
}
```

The result is
```
Dumping object(String)
 2020-10-15T18:32:40.5295578Z

Payload: {
  "ArbitraryString": "2020-10-15T18:32:40.5295578Z"
}
Decrypted payload: {
  "ArbitraryString": "2020-10-15T18:32:40.5295578Z"
}
Dumping object(String)
 10/15/2020 18:32:40
```

**Comment by rleverence-gd** [comment]
Hi @sushantmimani,

Yeah, that's expected. It doesn't try to reformat the date until after you take the JObject and convert it or the property itself to another object. It's a really weird quirk of how JObject is setup. When you do a `.ToString()` on the JObject itself, you're just returning the backing JSON without Newtonsoft.Json's interpretation of the data. That's actually how I'm working around the problem right now--I'm getting the decrtyped JObject, turning it back into the JSON, then again creating a JObject with the deserializer settings to make it have no datetime handling.

To see the problem first-hand, after you've decrypted the JObject, dump the `ArbitraryString` property as a string:
```c#
decryptedJObject.GetValue("ArbitraryString").ToString().Dump();
```

```
10/16/2020 14:54:28
```

**Comment by aka-bo** [comment]
This has been fixed in v0.1.6 (via #333)
